internal/pkg/docker/dockerfile/dockerfile.go

```diff
@@ -89,31 +118,60 @@ func (df *Dockerfile) parse() error {
 	file, err := df.fs.Open(df.path)
 
 	if err != nil {
-		return fmt.Errorf("read dockerfile: %w", err)
+		return fmt.Errorf("open Dockerfile: %w", err)
 	}
 	defer file.Close()
 
-	scanner := bufio.NewScanner(file)
-	parsedDockerfile := parseFromScanner(scanner)
+	f, err := afero.ReadFile(df.fs, file.Name())
+	if err != nil {
+		return fmt.Errorf("read Dockerfile %s error: %w", f, err)
+	}
+
+	parsedDockerfile, err := parseFromReader(string(f))
+	if err != nil {
+		return err
+	}
 
 	df.ExposedPorts = parsedDockerfile.ExposedPorts
+	df.HealthCheck = parsedDockerfile.HealthCheck
 	df.parsed = true
 	return nil
 }
 
-func parseFromScanner(scanner *bufio.Scanner) Dockerfile {
-	var line = ""
+// parseFromReader reads a Dockerfile and returns a Dockerfile struct.
+func parseFromReader(line string) (*Dockerfile, error) {
 	var df Dockerfile
 	df.ExposedPorts = []portConfig{}
-	for scanner.Scan() {
-		line = scanner.Text()
-		switch {
-		case strings.HasPrefix(line, "EXPOSE"):
-			currentPorts := parseExpose(line)
+
+	ast, err := parser.Parse(strings.NewReader(line))
+	if err != nil {
+		return nil, fmt.Errorf("parse reader: %w", err)
+	}
+
+	for _, child := range ast.AST.Children {
+		// ParseInstruction converts an AST to a typed instruction.
+		// Does prevalidation checks before parsing
+		// Example of an instruction is HEALTHCHECK CMD curl -f http://localhost/ || exit 1.
+		instruction, err := instructions.ParseInstruction(child)
+		if err != nil {
+			return nil, fmt.Errorf("parse instructions: %w", err)
+		}
+		inst := fmt.Sprint(instruction)
+
+		// Getting the value at a children will return the Dockerfile directive
+		switch d := child.Value; d {
```

Comment:
I really like what you did with `ParseInstruction`! I just realized that the instruction itself is already parsed into a `moby` type and we don't need to parse it again :D 

Can try something like this:
```go
var hc *HealthCheck
switch t := instruction.(type); t {
case *instructions.ExposeCommand:
// Read fields from t.
case *instructions.HealthCheckCommand:
    hc = &HealthCheck{
         interval: uint16(t.Health.Interval.Seconds()),
         // ... same for other fields
    }
}
```

Reply:
Talked about it offline and it provided a compiler error. Might have to tackle this later on.

Reply:
Sounds good but I think it's worthwhile addressing it because it will significantly remove code from us and will keep it in sync with how Docker parses the file.